### PR TITLE
allow patch to optionally create intermediate dirs

### PR DESCRIPTION
**Pull request by whyrusleeping**
Docker stuff is a headache without this option. We can also make it a different sub-command if that makes more sense.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: core/commands/object.go [hunk 1]
```diff
@@ -453,7 +453,9 @@ This removes the link named foo from the hash in $FOO_BAR and returns the
 resulting object hash.
 `,
 	},
-	Options: []cmds.Option{},
+	Options: []cmds.Option{
+		cmds.BoolOption("create", "p", "create intermediate directories on add-link"),
```
Comment (jbenet) [inline core/commands/object.go]: :+1:

File: core/commands/object.go [hunk 2]
```diff
@@ -619,12 +626,14 @@ func addLinkCaller(req cmds.Request, root *dag.Node) (key.Key, error) {
...
+	defer cancel()
 	childnd, err := ds.Get(ctx, childk)
 	if err != nil {
-		cancel()
 		return nil, err
 	}
-	cancel()
+
+	// ensure no link with that name already exists
+	_ = root.RemoveNodeLink(childname) // ignore error, only option is ErrNotFound
```
Comment (jbenet) [inline core/commands/object.go]: thx for the comment

File: core/commands/object.go [hunk 3]
```diff
@@ -638,31 +647,33 @@ func addLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname s
...
-	if err != nil {
-		return nil, err
-	}
-
-	ndprime, err := insertNodeAtPath(ctx, ds, nd, path[1:], toinsert)
-	if err != nil {
-		return nil, err
+		// if 'create' is true, we create directories on the way down as needed
+		if err == dag.ErrNotFound && create {
+			nd = &dag.Node{Data: ft.FolderPBData()}
```
Comment (jbenet) [inline core/commands/object.go]: whish this wasn't unixfs specific, but not sure what a good way is. ok for now.

File: core/commands/object.go [hunk 4]
```diff
@@ -638,31 +647,33 @@ func addLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname s
...
+		}
+	} else {
+		nd, err = child.GetNode(ctx, ds)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	err = root.RemoveNodeLink(path[0])
+	ndprime, err := insertNodeAtPath(ctx, ds, nd, path[1:], toinsert, create)
```
Comment (jbenet) [inline core/commands/object.go]: looks like this recursive construction works. but would love to see it tested more thoroughly.

File: test/sharness/t0051-object.sh [hunk 5]
```diff
@@ -145,6 +145,16 @@ test_object_cmd() {
...
 	'
+
+	test_expect_success "object patch --create works" '
+		OUT=$(ipfs object patch --create $EMPTY add-link a/b/c $FILE)
+	'
+
+	test_expect_success "result looks good" '
+		ipfs cat $OUT/a/b/c > p2_hwfile &&
+		test_cmp hwfile p2_hwfile
+	'
```
Comment (jbenet) [inline test/sharness/t0051-object.sh]: test edge + error cases too:

```
a
a/b
a/b/b/b/b
/ 
```
and one where the link already exists on the object.

recommend doing this sort of thing with a function:

```sh
test_patch_create_path() {
  root=$1
  name=$2
  target=$3

  test_expect_success "object patch --create works" ' ... '

  test_expect_success "result looks good" ' ... '

}

test_patch_create_path $EMPTY a/b/c $FILE

test_patch_create_path $EMPTY a $FILE

... 
```

**Comment by whyrusleeping** [comment]
dont merge quite yet, i might want to tweak something

File: core/commands/object.go (hunk 3 above)
Comment (whyrusleeping) [inline core/commands/object.go]: yeah, that was my argument when this option was originally brought up. we *could* just treat no data as a directory

**Comment by jbenet** [comment]
lmk when ready

**Comment by whyrusleeping** [comment]
@jbenet RFM

File: core/commands/object.go [hunk 6]
```diff
@@ -580,19 +586,59 @@ func rmLinkCaller(req cmds.Request, root *dag.Node) (key.Key, error) {
...
+
+func rmLink(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string) (*dag.Node, error) {
+	if len(path) == 1 {
+		// base case, remove node in question
+		err := root.RemoveNodeLink(path[0])
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = ds.Add(root)
```
Comment (jbenet) [inline core/commands/object.go]: this is just making sure the object is in `ds`? not sure why it's needed, in which case won't it be?

Comment (jbenet) [inline core/commands/object.go]: Oh, i see. if it's been modified!

Comment (jbenet) [inline core/commands/object.go]: these mutable nodes really bother me. i'd be so much more comfortable with:

node := dag.RemoveLink(root, path[0])

File: core/commands/object.go [hunk 7]
```diff
@@ -580,19 +586,59 @@ func rmLinkCaller(req cmds.Request, root *dag.Node) (key.Key, error) {
...
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	return newkey, nil
+	nd, err := nchild.GetNode(ctx, ds)
+	if err != nil {
+		return nil, err
+	}
```
Comment (jbenet) [inline core/commands/object.go]: this is a very common operation, maybe we want a func like:

```go
nd, err := root.GetChildNode(path[0])  // or dag.Child(root, path[0])
if err != nil {
  return nil, err
}
```

File: test/sharness/t0051-object.sh [hunk 8]
```diff
@@ -145,6 +161,23 @@ test_object_cmd() {
...
+
+	test_expect_success "output looks good" '
+		echo "QmZD3r9cZjzU8huNY2JS9TC6n8daDfT8TmE8zBSqG31Wvq" > multi_link_rm_exp &&
+		test_cmp multi_link_rm_out multi_link_rm_exp
+	'
+
+	test_patch_create_path $EMPTY a/b/c $FILE
+	test_patch_create_path $EMPTY a $FILE
+	test_patch_create_path $EMPTY a/b/b/b/b $FILE
```
Comment (jbenet) [inline test/sharness/t0051-object.sh]: i think we need to have newlines between these. i don't know why, but i've seen errors when I dont have them...

**Comment by jbenet** [comment]
@whyrusleeping comments above

File: test/sharness/t0051-object.sh (hunk 8 above)
Comment (whyrusleeping) [inline test/sharness/t0051-object.sh]: oh, really? i'll put them in

**Comment by whyrusleeping** [comment]
@jbenet RFM

File: merkledag/utils/utils.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+func AddLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname string, childk key.Key) (*dag.Node, error) {
+	if childname == "" {
+		return nil, errors.New("cannot create link with no name!")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
```
Comment (jbenet) [inline merkledag/utils/utils.go]: no timeout here. caller's responsibility. 

if you're worried about the "running out of time" problem (e.g. user gives context with timeout T, to do K steps, but we can estimate that if step 1 takes  > T/K time, we should fail early and avoid wasting resources on a doomed-to-run-out-of-time operation) use https://godoc.org/github.com/jbenet/go-context/frac

File: merkledag/utils/utils.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	childnd, err := ds.Get(ctx, childk)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure no link with that name already exists
+	_ = root.RemoveNodeLink(childname) // ignore error, only option is ErrNotFound
```
Comment (jbenet) [inline merkledag/utils/utils.go]: I checked and turns out `RemoveNodeLink` is broken. it only removes _one_ link. I think it should remove all.

File: merkledag/utils/utils.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+	err = root.AddNodeLinkClean(childname, childnd)
```
Comment (jbenet) [inline merkledag/utils/utils.go]: `Clean` is not a super clear name, maybe we should use:

```go
AddNodeLink(...) // yes ref
AddNodeLinkNoRef(...) // no ref

// or 

AddNodeLinkWithRef(...) // yes ref
AddNodeLink(...) // no ref
```

Comment (jbenet) [inline merkledag/utils/utils.go]: And, am still bothered by the mutability here. are you opposed to something like:

```go
root, err := root.AddNodeLink(childname, childnd)
```
based on
```go
l := append(l, elem)
```

File: merkledag/utils/utils.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+func InsertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string, toinsert key.Key, create bool) (*dag.Node, error) {
+	if len(path) == 1 {
+		return AddLink(ctx, ds, root, path[0], toinsert)
+	}
+
+	nd, err := root.GetLinkedNode(ctx, ds, path[0])
+	if err != nil {
+		// if 'create' is true, we create directories on the way down as needed
+		if err == dag.ErrNotFound && create {
+			nd = &dag.Node{Data: ft.FolderPBData()}
```
Comment (jbenet) [inline merkledag/utils/utils.go]: pass the default it as a parameter or something. otherwise we're baking _even more_ unix-fs specialness all over the place. this is not right.

File: merkledag/utils/utils.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	_, err = ds.Add(root)
+	if err != nil {
+		return nil, err
+	}
+	return root, nil
+}
+
+func InsertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string, toinsert key.Key, create bool) (*dag.Node, error) {
```
Comment (jbenet) [inline merkledag/utils/utils.go]: why not make path a ... "/" separated path? just because of the recursion?  it's probably a better UX to use paths. can use a helper function to do the recursion:

```go
func InsertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path string, toinsert key.Key, create bool) (*dag.Node, error) {
  p := strings.Split(path, "/")
  return insertNodeAtPath(ctx, ds, root, p, toinsert, create)
}
```

also, these functions with a bunch of params could use a helper type:

```go
type Editor struct {
  DS     dag.DAGService
  Root   dag.Node
  Create bool
}

func (de *Editor) InsertNodeAtPath(ctx context.Context, p []string, k key.Key) (*dag.Node, error) { ... }
```

The caller becomes clearer

```go
func main() {
  ed := dagutil.Editor{
    DS: dagService,
    Root: nd,
    Create: true,
  }

  if _, err := ed.InsertNodeAtPath(ctx, "a/b/c", c); err != nil { ... }
  if _, err := ed.InsertNodeAtPath(ctx, "a/b/d", d); err != nil { ... }
  ...
}
```

File: merkledag/utils/utils.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+			return nil, err
+		}
+	}
+
+	ndprime, err := InsertNodeAtPath(ctx, ds, nd, path[1:], toinsert, create)
+	if err != nil {
+		return nil, err
+	}
+
+	_ = root.RemoveNodeLink(path[0])
```
Comment (jbenet) [inline merkledag/utils/utils.go]: `RemoveNodeLink` still broken

File: merkledag/utils/utils_test.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	assertNodeAtPath(t, ds, root_b, path3, bk)
+
+	// now try overwriting a path
+	root_c, err := InsertNodeAtPath(context.Background(), ds, root_b, path, bk, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assertNodeAtPath(t, ds, root_c, path, bk)
+}
```
Comment (jbenet) [inline merkledag/utils/utils_test.go]: @whyrusleeping this is really not enough. these tests test ~4 of _many, many_ possible edge cases. they didn't even find the bug that i found by glancing (that `RemoveNodeLink` doesn't do what you expect it to).

take a look at the number of test cases i write for this sort of thing: https://github.com/jbenet/go-multiaddr/blob/master/multiaddr_test.go -- many of those uncovered bugs.

it usually helps to create functions to abstract out testing and allow expressing test cases as strings.

something like:

```go
root1 := testDag1(t)
root2 := testDag2(t)
root3 := testDag3(t)

testAdd(root1, "a", "aaaa", "<expected root hash>", "")
testAdd(root1, "a", "aaaa", "<expected root hash>", "") // different
testAdd(root1, "a", "aaaa", "<expected root hash>", "") // different
testAdd(root1, "b", "bbbb", "<expected root hash>", "") // removes existing links, checking all prior "b" are gone
testAdd(root1, "a/b/c", "ccccc", "<expected root hash>", "")
testAdd(root1, "a/b/c", "ccccc", "<expected root hash>", "")
testAdd(root1, "a/b/c/d/e/f/g", "ccccc", "<expected root hash>", "")
testAdd(root1, "a////b/c/d/e", "ccccc", "", "expected err str")
testAdd(root1, "a/b/c/d/////", "ccccc", "", "expected err str")
testAdd(root1, "/////", "ccccc", "", "expected err str")
```
and so on

File: merkledag/utils/utils.go (hunk 11 above)
Comment (whyrusleeping) [inline merkledag/utils/utils.go]: `l = append(l, elem)` will reuse l. it does not make any guarantees about returning a new array.

I guess we could make the node manipulation functions return a mutated copy, but that gets expensive when we start doing lots of dag operations. I would prefer to just have a mutable type of `Node` that retains existing behaviour, and maybe make `dag.Node` itself an interface with controlled creation.

In either case, thats not something that lies within the scope of this PR. We can file it as an issue and do it separately.

File: merkledag/utils/utils.go (hunk 9 above)
Comment (whyrusleeping) [inline merkledag/utils/utils.go]: mmm, my bad, thats a holdover from being in the commands lib with no explicit timeout.

File: merkledag/utils/utils.go (hunk 10 above)
Comment (whyrusleeping) [inline merkledag/utils/utils.go]: > I checked and turns out RemoveNodeLink is broken. 

mind sharing how you checked?

File: merkledag/utils/utils.go (hunk 12 above)
Comment (whyrusleeping) [inline merkledag/utils/utils.go]: hrm, how about instead of a bool for `create` we pass in a `dag.Node` and use the nil check as our flag?

```
if err == dag.ErrNotFound && template != nil {
    nd = template.Copy()
}
```

File: merkledag/utils/utils.go (hunk 13 above)
Comment (whyrusleeping) [inline merkledag/utils/utils.go]: so the editor would modify its internal 'root' variable after each call?

Comment (jbenet) [inline merkledag/utils/utils.go]: right.

File: merkledag/utils/utils.go (hunk 12 above)
Comment (jbenet) [inline merkledag/utils/utils.go]: works for me, and if it's in the template, then it's easy to do! could even pass a function: 

```go
type newNode func() *dag.Node
```

File: merkledag/utils/utils.go (hunk 10 above)
Comment (jbenet) [inline merkledag/utils/utils.go]: i read the code. 

```go
// Remove a link on this node by the given name
func (n *Node) RemoveNodeLink(name string) error {
	n.encoded = nil
	for i, l := range n.Links {
		if l.Name == name {
			n.Links = append(n.Links[:i], n.Links[i+1:]...)
			return nil   // <---- means it only does it **once**
		}
	}
	return ErrNotFound
}
```

File: merkledag/utils/utils.go (hunk 11 above)
Comment (jbenet) [inline merkledag/utils/utils.go]: > `l = append(l, elem)` will reuse l. it does not make any guarantees about returning a new array.

derp.

> In either case, thats not something that lies within the scope of this PR. We can file it as an issue and do it separately.

sounds good

File: merkledag/utils/utils.go (hunk 13 above)
Comment (whyrusleeping) [inline merkledag/utils/utils.go]: @jbenet is this something you want in this PR? sure it would be nice, but its a bit much for a PR about adding an extra flag to object patch.

Comment (whyrusleeping) [inline merkledag/utils/utils.go]: > why not make path a ... "/" separated path?

because paths separated by a given delimiter are a unixfs thing, merkledag traversal is just a series of links, whose names can contain (technically) anything. It makes the most sense to me to have it be an array of names.

Comment (jbenet) [inline merkledag/utils/utils.go]: > paths separated by a given delimiter are a unixfs thing

no, this is part of the dag. path traversal is why ipfs has human readable string names in the links.

> @jbenet is this something you want in this PR? sure it would be nice, but its a bit much for a PR about adding an extra flag to object patch.

yes because this is what patch is, this data structure is what the patch command is doing. it will make every other patch command easier to reason about and call. it will make things like the write-and-update-merkle-root part of ipns fs easier to liberate and use with just ipfs objects (not ipns).

File: merkledag/node.go [hunk 16]
```diff
@@ -129,13 +129,23 @@ func (n *Node) AddRawLink(name string, l *Link) error {
...
 		}
 	}
-	return ErrNotFound
+	n.Links = good
+
+	if !found {
+		return ErrNotFound
+	}
+
+	return nil
```
Comment (jbenet) [inline merkledag/node.go]: @whyrusleeping pls add test for this o/ else it may get broken again

File: merkledag/utils/utils.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+)
+
+type Editor struct {
+	root *dag.Node
+	ds   dag.DAGService
+}
```
Comment (jbenet) [inline merkledag/utils/utils.go]: this looks great!

**Comment by jbenet** [comment]
@whyrusleeping LGTM -- one minor thing: test the Remove Link func in node. other than that, RFM

**Comment by whyrusleeping** [comment]
@jbenet theres those tests, I'll have more merkledag test coverage on 0.4.0 as i use and test FetchGraph

**Comment by jbenet** [comment]
thanks @whyrusleeping :heart: :clap: :heart: :clap: :heart: :clap: :heart: :clap: :heart: :clap:
